pkg/sql/explain_bundle.go

```diff
@@ -0,0 +1,199 @@
+package sql
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/internal/client"
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/sql/opt/memo"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+)
+
+// handleExplainBundle creates the bundle and returns the result for an
+// EXPLAIN BUNDLE statement. The result is text containing a URL for the bundle.
+func handleExplainBundle(
+	ctx context.Context, res RestrictedCommandResult, plan *planTop, execCfg *ExecutorConfig,
+) error {
+	res.ResetStmtType(&tree.ExplainBundle{})
+	res.SetColumns(ctx, sqlbase.ExplainBundleColumns)
+
+	id, err := buildStatementBundle(ctx, plan, execCfg)
+	if err != nil {
+		return err
+	}
+
+	url := fmt.Sprintf("%s/_admin/v1/stmtbundle/%d", execCfg.AdminURL(), id)
+	text := fmt.Sprintf(""+
+		"Download the bundle from:\n"+
+		"  %s\n"+
+		"or from the Admin UI (Advanced Debug -> Statement Diagnostics).",
+		url,
+	)
+	return res.AddRow(ctx, tree.Datums{tree.NewDString(text)})
+}
+
+// buildStatementBundle collects metadata related the planning and execution of
+// the statement, generates a bundle, stores it in the
+// system.statement_bundle_chunks table and adds an entry in
+// system.statement_diagnostics.
+//
+// Returns the bundle ID, which is the key for the row added in
+// statement_diagnostics.
+func buildStatementBundle(
+	ctx context.Context, plan *planTop, execCfg *ExecutorConfig,
+) (bundleID int64, _ error) {
+	b := makeStmtBundleBuilder(plan)
+
+	b.addStatement()
+	b.addOptPlans()
+	b.addExecPlan()
+
+	buf, err := b.finalize()
+	if err != nil {
+		return 0, err
+	}
+
+	db, ie := execCfg.DB, execCfg.InternalExecutor
+	fingerprint := tree.AsStringWithFlags(plan.stmt.AST, tree.FmtHideConstants)
+	statement := tree.AsString(plan.stmt.AST)
+	description := "query support bundle XXX more info here"
+
+	err = db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
+		// Insert the bundle into system.statement_bundle_chunks.
+		// TODO(radu): split in chunks.
+		row, err := ie.QueryRowEx(
+			ctx, "statement-bundle-chunks-insert", txn,
+			sqlbase.InternalExecutorSessionDataOverride{User: security.RootUser},
+			"INSERT INTO system.statement_bundle_chunks(description, data) VALUES ($1, $2) RETURNING id",
+			description,
+			tree.NewDBytes(tree.DBytes(buf.String())),
+		)
+		if err != nil {
+			return err
+		}
+		chunkID := row[0].(*tree.DInt)
+		chunks := tree.NewDArray(types.Int)
+		if err := chunks.Append(chunkID); err != nil {
+			return err
+		}
+
+		row, err = ie.QueryRowEx(
+			ctx, "statement-bundle-info-insert", txn,
+			sqlbase.InternalExecutorSessionDataOverride{User: security.RootUser},
+			`INSERT INTO
+			  system.statement_diagnostics(statement_fingerprint, statement, collected_at, bundle_chunks)
```

Comment:
@RaduBerinde just to think through how the UI would show this, we would query the `statement_diagnostics` table by fingerprint, correct?